File: Sources/HedosCLI/Commands/Image.swift
import ArgumentParser
import Foundation
import HedosKernel

struct Image: AsyncParsableCommand {
    static let configuration = CommandConfiguration(
        commandName: "image",
        abstract: "Generate an image from a prompt and write it to a file.")

    @OptionGroup var global: GlobalOptions

    @Argument(help: "Model id or name.")
    var model: String

    @Argument(help: "The image prompt.")
    var prompt: String

    @Option(name: .long, help: "Number of denoising steps.")
    var steps: Int?

    @Option(name: .long, help: "Random seed.")
    var seed: Int?

    @Option(name: [.customShort("o"), .long], help: "Output file path (.png).")
    var output: String?

    func run() async throws {
        let kernel = Session.kernel()
        let shelf = try await Session.shelf(kernel)
        let record = try Session.resolve(model, in: shelf, capability: .image)

        var fields: [String: JSONValue] = ["prompt": .string(prompt)]
        if let steps { fields["steps"] = .int(steps) }
        if let seed { fields["seed"] = .int(seed) }

        let jobID = try await kernel.submit(record.id, .image, payload: .object(fields))

        var artifactIDs: [String] = []
        var failure: String?
        loop: for await event in await kernel.scheduler.events(id: jobID) {
            switch event {
            case .status(let status):
                if !global.json { Out.err(status) }
            case .progress(let progress):
                if !global.json, let step = progress.step, let total = progress.totalSteps {
                    Out.err("step \(step)/\(total)")
                }
            case .done(let ids):
                artifactIDs = ids
                break loop
            case .failed(let message):
                failure = message
                break loop
            case .cancelled:
                failure = "generation cancelled."
                break loop
            default:
                break
            }
        }

        if let failure { throw CLIError(failure) }
        guard let first = artifactIDs.first,
            let stored = try await kernel.artifactStore.url(id: first)
        else { throw CLIError("\(record.displayName) produced no image.") }

        let finalPath = try deliver(stored, to: output)
        if global.json {
            try Out.json(MediaReport(model: record.id, path: finalPath))
        } else {
            Out.line(finalPath)
        }
    }
}

File: cli/src/commands/image.rs
//! `hedos image <model> <prompt>` — generate an image and write a PNG file.

use std::collections::BTreeMap;
use std::path::PathBuf;

use clap::Args;
use kernel::jobs::JobEvent;
use kernel::records::{Capability, JsonValue};

use crate::error::CliError;
use crate::support::output::Out;
use crate::support::session::{self, Session};

/// Arguments for `image`.
#[derive(Args)]
pub struct ImageArgs {
    /// The image model (name, alias, or id).
    model: String,
    /// The prompt to render.
    prompt: String,
    /// Number of diffusion steps.
    #[arg(long)]
    steps: Option<i64>,
    /// The random seed.
    #[arg(long)]
    seed: Option<i64>,
    /// Where to write the `.png` (default: `<model>.png` in the current dir).
    #[arg(short, long)]
    output: Option<PathBuf>,
}

/// Run the `image` command.
pub async fn run(args: ImageArgs, out: &Out) -> Result<(), CliError> {
    let session = Session::open()?;
    let shelf = session.shelf_or_discover().await?;
    let record = session::resolve(&args.model, &shelf, Some(&Capability::image()))?;

    let payload = image_payload(&args.prompt, args.steps, args.seed);
    let job_id = session
        .kernel
        .submit(&record.id, Capability::image(), payload)
        .await?;
    let mut events = session.kernel.scheduler().events(&job_id);

    let mut artifacts: Vec<String> = Vec::new();
    while let Some(event) = events.recv().await {
        match event {
            JobEvent::Progress(progress) => {
                out.progress(&format!("{}%", (progress.fraction * 100.0) as i64));
            }
            JobEvent::Status(status) => out.progress(&status),
            JobEvent::Done { result } => {
                artifacts = result;
                break;
            }
            JobEvent::Failed { message } => {
                out.progress_done();
                return Err(CliError::new(message));
            }
            JobEvent::Cancelled => {
                out.progress_done();
                return Err(CliError::new("generation was cancelled"));
            }
            _ => {}
        }
    }
    out.progress_done();

    let artifact = artifacts
        .first()
        .ok_or_else(|| CliError::new(format!("{} produced no image", record.display_name())))?;
    let bytes = session
        .kernel
        .artifact_data(artifact)
        .await?
        .ok_or_else(|| CliError::new("the image artifact is missing"))?;
    let path = args
        .output
        .unwrap_or_else(|| crate::support::paths::default_path(record.display_name(), "png"));
    std::fs::write(&path, bytes)?;

    out.line(&format!("wrote {}", path.display()));
    out.json(&serde_json::json!({
        "model": record.id,
        "path": path.display().to_string(),
    }));
    Ok(())
}

/// An `image` payload: the prompt plus optional steps and seed.
fn image_payload(prompt: &str, steps: Option<i64>, seed: Option<i64>) -> JsonValue {
    let mut payload = BTreeMap::new();
    payload.insert("prompt".to_owned(), JsonValue::String(prompt.to_owned()));
    if let Some(steps) = steps {
        payload.insert("steps".to_owned(), JsonValue::Int(steps));
    }
    if let Some(seed) = seed {
        payload.insert("seed".to_owned(), JsonValue::Int(seed));
    }
    JsonValue::Object(payload)
}
